### PR TITLE
Go 1.8 compatibility (part 2)

### PR DESCRIPTION
**Pull request by aarzilli**
Do not commit / Do not review.

### CODE REVIEW

**Comment by aarzilli** [comment]
Partially blocked by https://github.com/golang/go/issues/17830

**Comment by aarzilli** [comment]
The go issue has been fixed, this can be reviewed now.

File: dwarf/line/state_machine.go [hunk 1]
```diff
@@ -151,9 +161,10 @@ func execSpecialOpcode(sm *StateMachine, instr byte) {
 
 	sm.lastDelta = int(sm.dbl.Prologue.LineBase + int8(decoded%sm.dbl.Prologue.LineRange))
 	sm.line += sm.lastDelta
-	sm.address += uint64(decoded / sm.dbl.Prologue.LineRange)
+	sm.address += uint64(decoded/sm.dbl.Prologue.LineRange) * uint64(sm.dbl.Prologue.MinInstrLength)
```
Comment (dr2chase) [inline dwarf/line/state_machine.go]: I'm doubtful about this change.  I ended up with this, instead:
```
                sm.isStmt = true
        }
 
-       sm.lastDelta = int(sm.dbl.Prologue.LineBase + int8(decoded%sm.dbl.Prologue.LineRange))
+       var deltapc uint64
+       var deltaline int
+
+       switch opcode {
+       case 249:
+               deltaline = 5
+               deltapc = 23
+       case 255:
+               deltaline = 1
+               deltapc = 24
+       default:
+               deltaline = int(sm.dbl.Prologue.LineBase + int8(decoded%sm.dbl.Prologue.LineRange))
+               deltapc = uint64(decoded / sm.dbl.Prologue.LineRange)
+       }
+
+       sm.lastDelta = deltaline
        sm.line += sm.lastDelta
-       sm.address += uint64(decoded / sm.dbl.Prologue.LineRange)
+       sm.address += deltapc
        sm.basicBlock = false
        sm.lastWasStandard = false
```
I got this from https://go.googlesource.com/go/+/f9feaffdf51967bc5cf3c9363db9ddee98c9b3a0/src/cmd/link/internal/ld/dwarf.go#918

```
		} else {
			// Use opcode 249 (pc+=23, lc+=5) or 255 (pc+=24, lc+=1).
			//
```
and I don't see any way that the change proposed here takes us from a state where the PC increments were not 23 and 24 to one where they are (because 23 and 24 have no common factors other than 1).

Comment (aarzilli) [inline dwarf/line/state_machine.go]: This change is just a formality MinInstrLen, at least on amd64, is always 1. AFAIK delve always interpreted special opcodes correctly:

```
objdump --dwarf ./testvariables2 | grep -A11 'debug_line' 
Raw dump of debug contents of section .debug_line:

  Offset:                      0x0
  Length:                      76002
  DWARF Version:               2
  Prologue Length:             7461
  Minimum Instruction Length:  1
  Initial value of 'is_stmt':  1
  Line Base:                   -4
  Line Range:                  10
  Opcode Base:                 10

```

plugging these values into that code we get https://play.golang.org/p/PhTBbfJ8GX :

```249 = addr 23 line 5
255 = addr 24 line 1
```

I don't know why link would need a special branch to handle 249 and 255. What we interpreted wrong was the standard opcode DW_LNS_const_add_pc, which wasn't emitted (or was only emitted rarely) by previous versions of link.

File: dwarf/line/state_machine.go [hunk 2]
```diff
@@ -26,6 +26,11 @@ type StateMachine struct {
 	endSeq          bool
 	lastWasStandard bool
 	lastDelta       int
+	// if valid is true the current value of the state machine is the addres of
```
Comment (derekparker) [inline dwarf/line/state_machine.go]: nit: s/addres/address/

Comment (aarzilli) [inline dwarf/line/state_machine.go]: Done.

File: proc/proc_test.go [hunk 3]
```diff
@@ -2374,8 +2373,25 @@ func TestIssue683(t *testing.T) {
 				break
 			}
 		}
-		if !goterror {
-			t.Fatal("expeceted an error we didn't get")
+		ver, _ := ParseVersionString(runtime.Version())
+
+		if !(ver.Major < 0 || ver.AfterOrEqual(GoVersion{1, 8, -1, 0, 0})) {
```
Comment (derekparker) [inline proc/proc_test.go]: Would be nice to have consts that represent the different Go versions for simplicity.

File: dwarf/line/line_parser_test.go [hunk 4]
```diff
@@ -67,11 +67,15 @@ func TestDebugLinePrologueParser(t *testing.T) {
 		t.Fatal("Initial value of 'is_stmt' not parsed correctly", prologue.InitialIsStmt)
 	}
 
-	if prologue.LineBase != int8(-1) {
+	if prologue.LineBase != int8(-1) && prologue.LineBase != int8(-4) {
```
Comment (derekparker) [inline dwarf/line/line_parser_test.go]: nit: would be nice to use consts for these values `int8(-1)` and `int8(-4)`.

File: dwarf/line/line_parser_test.go [hunk 5]
```diff
@@ -67,11 +67,15 @@ func TestDebugLinePrologueParser(t *testing.T) {
...
+		// go < 1.8 uses -1
+		// go >= 1.8 uses -4
 		t.Fatal("Line base not parsed correctly", prologue.LineBase)
 	}
 
-	if prologue.LineRange != uint8(4) {
+	if prologue.LineRange != uint8(4) && prologue.LineRange != uint8(10) {
```
Comment (derekparker) [inline dwarf/line/line_parser_test.go]: nit: same as above.

File: dwarf/line/state_machine.go [hunk 6]
```diff
@@ -27,6 +27,11 @@ type StateMachine struct {
 	endSeq          bool
 	lastWasStandard bool
 	lastDelta       int
+	// if valid is true the current value of the state machine is the address of
```
Comment (derekparker) [inline dwarf/line/state_machine.go]: nit: s/if valid is true/valid is true if/

File: service/test/integration2_test.go [hunk 7]
```diff
@@ -1048,8 +1048,16 @@ func TestSkipPrologue2(t *testing.T) {
 
 		callme3 := findLocationHelper(t, c, "main.callme3", false, 1, 0)[0]
 		callme3Z := findLocationHelper(t, c, "main.callme3:0", false, 1, 0)[0]
-		// callme3 does not have local variables therefore the first line of the function is immediately after the prologue
-		findLocationHelper(t, c, "callme.go:18", false, 1, callme3Z)
+		ver, _ := proc.ParseVersionString(runtime.Version())
```
Comment (derekparker) [inline service/test/integration2_test.go]: Fail test on error here.

File: service/test/integration2_test.go [hunk 8]
```diff
@@ -1048,8 +1048,16 @@ func TestSkipPrologue2(t *testing.T) {
...
+		if ver.Major < 0 || ver.AfterOrEqual(proc.GoVersion{1, 8, -1, 0, 0}) {
```
Comment (derekparker) [inline service/test/integration2_test.go]: nit: Use const to represent this version (`proc.GoVersion{1, 8, -1, 0, 0}`).

File: proc/proc.go [hunk 9]
```diff
@@ -359,10 +359,18 @@ func (dbp *Process) Continue() error {
 		case dbp.CurrentThread.CurrentBreakpoint == nil:
 			// runtime.Breakpoint or manual stop
 			if dbp.CurrentThread.onRuntimeBreakpoint() {
-				for i := 0; i < 2; i++ {
+				// Single-step current thread until we exit runtime.breakpoint and
+				// runtime.Breakpoint.
+				// On go < 1.8 it was sufficient to single-step twice on go1.8 a change
+				// to the compiler requires 4 steps
```
Comment (derekparker) [inline proc/proc.go]: nit: Add punctuation at end of sentence (sorry).

File: proc/proc.go [hunk 10]
```diff
@@ -359,10 +359,18 @@ func (dbp *Process) Continue() error {
...
+				for {
 					if err = dbp.CurrentThread.StepInstruction(); err != nil {
 						return err
 					}
+					loc, err := dbp.CurrentThread.Location()
+					if err != nil || loc.Fn == nil || (loc.Fn.Name != "runtime.breakpoint" && loc.Fn.Name != "runtime.Breakpoint") {
```
Comment (derekparker) [inline proc/proc.go]: nit: (don't feel strongly, can ignore) might be nice to wrap this logic in a function like `isInRuntimeBreakpoint`.

File: proc/proc_test.go [hunk 11]
```diff
@@ -2374,8 +2373,25 @@ func TestIssue683(t *testing.T) {
...
+			// in go1.8 the debug_line entry is correct for this piece of code
+			if !goterror {
```
Comment (aarzilli) [inline proc/proc_test.go]: I think it's better if I remove the error check completely, whether or not we get the error is irrelevant and depends on what the compiler does, what's important is that the test doesn't panic.

File: proc/proc_test.go (hunk 3 above)
Comment (aarzilli) [inline proc/proc_test.go]: Will do in a separate PR since it doesn't have anything to do with 1.8 compatibility

File: service/test/integration2_test.go (hunk 7 above)
Comment (aarzilli) [inline service/test/integration2_test.go]: Second value is not an error, it's a bool indicating whether the version string is parsable. In particular we can't parse the version string of development versions of go (because they contain the sha-1 abbrev of HEAD), this condition is covered by the `ver.Major < 0` check.

File: proc/proc.go (hunk 10 above)
Comment (aarzilli) [inline proc/proc.go]: it's short and only used here, imho it's better inline.

**Comment by aarzilli** [comment]
Done.

**Comment by ignatov** [comment]
@derekparker Do you have any estimates for that fix?

File: service/test/integration2_test.go (hunk 7 above)
Comment (derekparker) [inline service/test/integration2_test.go]: SGTM
